### PR TITLE
feat: [M12] SERVE-3: make session rewind reachable — RewindTree has no caller

### LINKED ISSUE #305: [M12] SERVE-3: make session rewind reachable — RewindTree has no caller
Found by `/closure-audit` (2026-08-18), pass A. Capability row `SERVE-3` in
`docs/feature-matrix.md`.

## Capability
A person can rewind a generation session to a previous turn boundary and branch a new
continuation from it, from an actual entry point.

## The gap

`src/serve/rewind.py`'s `RewindTree` is built, unit-tested (`tests/test_serve.py`) and described
in three separate design documents:

- `docs/usage.md:273-274`
- `docs/design/01-architecture-seam.md:66`
- `docs/design/14-inference-engine.md:107-108`

It has **zero non-test importers** anywhere in `src/` or `scripts/` (105 src modules and 45
scripts enumerated; `tests/` deliberately excluded, since a test is not a consumer). The closest
thing to a call site is `src/serve/sessions.py:21,160-163`, which mentions
`"hand to serve.rewind.RewindTree.commit(...)"` **in a docstring** — prose, not a call.
`scripts/generate.py` does not mention rewind at all.

So the layer closed and the capability never reached a person. It is also load-bearing beyond
serving: #202 names `RewindTree`'s state snapshot/restore as the mechanism that would let the
hybrid attention+SSM arm roll back by true state restoration instead of re-prefill, which is the
only thing that arm exists to test.

## Scope

In scope:
- Wire `RewindTree` into `src/serve/sessions.py` for real — the commit/rewind path the docstring
  already describes.
- Expose it through one entry point a person can actually use: flags on `scripts/generate.py`
  (commit a turn, rewind N turns, branch and continue).
- Cover the wired path with a test that goes through the entry point, not just the class.

Out of scope:
- The Swift engine side.
- #202's hybrid-arm rollback experiment — this unblocks it, it is not this.
- Any change to `RewindTree`'s own semantics; it is tested and works. This is about reaching it.

## Definition of done
- [x] `src/serve/sessions.py` really calls `RewindTree`; the `:160-163` docstring either describes
      code that now exists or is rewritten to stop describing code that does not.
- [x] `scripts/generate.py` exposes commit / rewind / branch, and `--help` documents them.
- [x] Rewinding restores the fixed-size recurrent state — a rewound-then-regenerated continuation
      is identical to generating it directly from that prefix under the same seed.
- [x] A test drives the flow through the entry point end to end, not through `RewindTree` alone.
- [x] `docs/usage.md` shows the actual command, replacing the current description of a capability
      with no entry point.
- [x] `docs/feature-matrix.md` `SERVE-3` moves off `Planned` with its per-layer cells filled in;
      `docs/test-plan.md`'s `SERVE-3` Gaps cell is rewritten.
- [x] No dead paths left: if any part of `RewindTree` stays unreachable after this, say which and
      why.

## Edge cases
- Rewinding further than the number of committed turns → a clear error, not an exception from
  inside the tree.
- Rewinding to turn 0 / an empty session.
- Branching twice from the same node, then rewinding across the branch point.
- A session whose state was never committed → rewind is a no-op with a stated message.
- Memory: an unbounded tree over a long session — state whether nodes are pruned, and if not, say so.

## Verification
- `.venv/bin/python -m pytest tests/test_serve.py -q`
- `.venv/bin/python scripts/generate.py --config config/toy.yaml ...` exercising commit → rewind →
  branch, with the identity check above.
- `scripts/smoke_test.py --data <fresh toy split>` still green.

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #305

## Summary

`RewindTree` was built, unit-tested and described in three design documents, but had **zero
non-test importers** in `src/` or `scripts/`. The layer closed; the capability never reached a
person. This wires it end to end through one entry point.

- **`src/serve/sessions.py` — `SessionHistory`, the composition point.** `commit_turn` is
  `tree.commit(store.get_state(sid))`; `rewind_to` is `store.set_state(sid, tree.rewind(id))`.
  Portable, numpy-free, opaque `State` only. Guards **reject and name the available depth** —
  none clamps to the root, none silently no-ops. The docstring states that rewinds count
  **retained** boundaries, not wall-clock turns, because `RewindTree._detach` reparents an
  evicted node's children onto its grandparent.
- **`scripts/generate.py --interactive` — the entry point.** A third mode alongside `--prompt`
  and `--chat`: one stateful session for the whole run, `/rewind [n|#id]`, `/tree`, `/help`,
  `/quit`, sized by `--rewind-depth` (default 32; `0` disables rewind and builds no tree;
  negative is an argparse error). The loop is extracted to a module-level `interactive_repl(...)`
  with `run_turn` / `read_line` / `emit` injected, so tests drive it without a pty and without a
  backend. `--help` documents every command via an `epilog`.
- **`src/serve/generate.py` — additive `prefill: bool = True`.** The one net-new line of library
  work: `SessionStore.prefill` is fresh-session-only, and a session restored from a snapshot has
  an unknowable position above the seam, so turns after the first (and every turn after a rewind)
  step the prompt in instead. The default path is byte-identical to before.
- **`--chat` is deliberately left untouched.** It is stateless by construction — a fresh session
  per turn, whole transcript re-rendered — so bolting `/rewind` onto it would degenerate to
  popping a `messages` list: the same "no real caller" defect in a new costume.

## Criterion 3, proven by output — and watched failing first

The whole defect being fixed is a class with no caller, so "has a caller" is trivial to fake. The
proof is an **output identity** against a fake model whose logits depend on the recurrent state:

```
--- baseline: StatefulFakeModel, rewind intact
    out_B='MKGIMP'  out_X='GCFOLO'  out_B'='MKGIMP'
    half 1 (out_B' == out_B) : PASS      half 2 (out_B' != out_X) : PASS
--- MUTATION A: rewind_to never calls store.set_state
    out_B='MKGIMP'  out_X='GCFOLO'  out_B'='CLAAIJ'
    half 1 (out_B' == out_B) : FAIL      half 2 (out_B' != out_X) : PASS
--- MUTATION B: rewind_to restores a fresh zeroed state
    out_B='MKGIMP'  out_X='GCFOLO'  out_B'='HDBCLN'
    half 1 (out_B' == out_B) : FAIL      half 2 (out_B' != out_X) : PASS
--- MUTATION C: state-independent FakeModel, rewind intact
    out_B='CCCCCC'  out_X='CCCCCC'  out_B'='CCCCCC'
    half 1 (out_B' == out_B) : PASS      half 2 (out_B' != out_X) : FAIL
```

Read honestly: **half 1 is what catches an unrestored state** (mutations A and B). **Half 2 is
what catches a vacuous test** — it fails only under mutation C, where the model's logits do not
depend on the state, which is exactly the trap `tests/test_serve.py`'s existing `FakeModel`
(`logits = eye(V)[token % V]`) would have walked the test into. That case is pinned permanently as
`test_state_independent_model_makes_the_identity_vacuous`, so the reason the stateful model is
required cannot quietly rot.

`StatefulFakeModel` uses a rolling hash rather than the running sum the plan sketched: greedy
decoding on `eye(V)[sum % V]` has absorbing fixed points, so distinct prefixes converge to the
same repeated token within a few steps and half 2 goes vacuous again.

## The over-deep guard, watched rejecting by hand

Verification step 5, real CLI, `config/toy.yaml`, random init, greedy — exact stdout:

```
rewind: up to 8 retained turn boundaries, at most 155648 bytes (8 x 19456 B/snapshot).
interactive mode — type text to continue the session; /help for commands.
>>> tttttttttttttttttttt[committed #1; depth 2, 2/8 retained]
>>>   #0  parent=None  children=[1]  '(start)'
* #1  parent=0  children=[]  'the cattttttttttttttttttttt'
>>> nnnnnnnnnnnnnnnnnnnn[committed #2; depth 3, 3/8 retained]
>>> rewound to #1 (depth 2); transcript:
  #0: '(start)'
  #1: 'the cattttttttttttttttttttt'
>>> yyyyyyyyyyyyyyyyyyyy[committed #3; depth 3, 4/8 retained]
>>>   #0  parent=None  children=[1]  '(start)'
  #1  parent=0  children=[2, 3]  'the cattttttttttttttttttttt'
  #2  parent=1  children=[]  'sat downnnnnnnnnnnnnnnnnnnnn'
* #3  parent=1  children=[]  'ran awayyyyyyyyyyyyyyyyyyyyy'
>>> error: cannot rewind 99 turn(s): session 'repl' retains 3 boundary/boundaries including the current one, so at most 2 rewind hop(s) are possible (retained ids: [0, 1, 2, 3])
>>> error: unknown command '/frobnicate' — /help lists the commands. It was NOT sent to the model.
```

The post-rewind continuation (`yyy…`) differs from the discarded one (`nnn…`); `/tree` shows node
`#1` with two children; `/rewind 99` prints the available depth with **no traceback and no change
to the session**; `/frobnicate` is rejected rather than fed to the model.

## No dead paths (DoD 10)

Every `RewindTree` member, and whether the entry point reaches it. Backed by
`grep -nF "self.tree." src/serve/sessions.py`:

| Member | Reachable from `--interactive`? | Where |
|---|---|---|
| `commit` | yes | `SessionHistory.commit_turn` (`sessions.py:250`) ← REPL startup + every turn |
| `rewind` | yes | `SessionHistory.rewind_to` (`:284`) ← `/rewind` |
| `current` | yes | `:263`, `:294`, `:302` ← `/tree`'s `*` marker, `depth()` |
| `parent` | yes | `:270`, `:306`, `:311` ← `rewind_turns` hops, `depth()`, `/tree` |
| `children` | yes | `:311` ← `/tree` |
| `__contains__` | yes | `:298` (`retained_ids`) ← `/tree`, every guard message |
| `__len__` | yes | `SessionHistory.__len__` (`:294`) ← the `[committed …]` and `/tree` status lines |
| `max_depth` | yes | `:315` ← startup budget line, `/tree` header |
| `_evict`, `_pick_victim`, `_detach` | **no — internal only** | called by `commit` when the LRU cap is exceeded. They are reached *through* the entry point at `--rewind-depth k` once k+1 turns are committed (covered by `test_eviction_bounds_the_tree_and_a_deeper_rewind_errors`), but never called by name outside `rewind.py` and its unit tests. That is correct: they are the cap's implementation, not API. |

Nothing in `src/serve/rewind.py` is now unreachable.

## Feature-matrix convention note

`SERVE-3` is written `done | done | none | none | Shipped`. CUDA is `none` because the path is
backend-agnostic but is exercised on no CUDA host and by no CUDA test; Swift is `none` (out of
scope, and the engine exists, so not `n/a`). **Shipped-with-`none` precedent checked and found:**
`LSP-10` (`docs/feature-matrix.md:123`, #279) is `Shipped` with `none` in the Swift column. The
convention is not invented here.

## Testing

- [x] Tests added/updated — `tests/test_rewind_entry_point.py` (19 new, portable, every assertion
      through `interactive_repl`; `grep -c RewindTree` = **0**), `tests/test_serve.py` (+14
      `SessionHistory` unit tests and 1 mlx-gated toy-scale opaque-state test),
      `tests/test_generate.py` (+3 `prefill=False` equivalence tests)
- [x] All tests passing — `.venv/bin/python -m pytest -q -rs` → **1722 passed, 22 skipped** in
      378.86 s
- [x] Untouched gates pass **unedited** — `tests/test_import_guard.py` and
      `tests/test_workflow_triggers.py`, 12 passed (no CI job added, no new portable module)
- [x] Smoke gate green on a **fresh** toy split — `SMOKE TEST PASSED ✅ resume is exact and eval runs.`
- [x] Manual testing completed — verification step 5 above, plus `--help` printing all three
      commands and both flags

**macOS CI cost (#315 constraint).** The one mlx-gated addition is toy-scale and measured:
`pytest tests/test_serve.py -q -k "mlx and rewind"` → **1 passed in 0.13 s** (0.45 s wall
including interpreter startup). Well under the 10 s bound; nothing slow was added to the macOS PR
path. `tests/test_rewind_entry_point.py` is backend-free and runs in the Linux `portable` job
(19 tests in 0.08 s).

## Parked

1 finding appended to `docs/parked-findings.md` (`RewindTree` has no public retained-id
enumeration, so `SessionHistory` duplicates the bookkeeping). Not acted on.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_011ur4BdaFTqnwqSUAx5TrX8

### CODE REVIEW

File: scripts/generate.py [hunk 1]
```diff
@@ -24,23 +32,193 @@
...
+            continue
+
+        if not text.startswith("/"):
+            out = run_turn(text, prefill=fresh)
+            fresh = False
+            if rewind_enabled:
+                node_id = history.commit_turn()
+                transcript[node_id] = text + out
+                emit(f"[committed #{node_id}; depth {history.depth()}, "
+                     f"{len(history)}/{history.max_depth} retained]")
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply scripts/generate.py:159]: ✅ Fixed in 5611219 — `transcript` in `interactive_repl` is now pruned to the tree's retained ids on every `commit_turn()`, so it stays bounded by `--rewind-depth` instead of growing without bound over a long session.

File: scripts/generate.py [hunk 2]
```diff
@@ -24,23 +32,193 @@
...
+                continue
+            if len(command_args) > 1:
+                emit(f"error: /rewind takes at most one argument; got {command_args}")
+                continue
+            try:
+                count, absolute = _parse_rewind_arg(
+                    command_args[0] if command_args else "1")
+                if absolute:
+                    target = history.rewind_to(count)
+                elif count >= 1 and history.depth() <= 1:
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply scripts/generate.py]: ✅ Fixed in 5611219 — the branch now only special-cases the default `/rewind` (count 1) at the root for the friendly 'session is empty' message. `/rewind 2` (or any count > 1) at the root now falls through to `rewind_turns`, which raises the precise 'cannot rewind N turn(s) ... at most 0 rewind hop(s) are possible' error naming the actual requested count. Added `test_rewind_multiple_at_root_reports_over_deep_not_empty` to lock this in.

File: src/serve/sessions.py [hunk 3]
```diff
@@ -191,3 +206,127 @@ def _maybe_evict(self) -> list[str]:
...
+        # `retained_ids` filters it through the tree's own membership test rather than
+        # reaching into `RewindTree._nodes`.
+        self._committed: list[int] = []
+
+    # --- the composition ---
+    def commit_turn(self) -> int:
+        """Snapshot the live session as a turn boundary. Returns the new node id."""
+        node_id = self.tree.commit(self.store.get_state(self.session_id))
+        self._committed.append(node_id)
+        return node_id
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/serve/sessions.py:258]: ✅ Fixed in 5611219 — `SessionHistory._committed` is now pruned to `retained_ids()` on every `commit_turn()`, so the list itself stays bounded by the tree's LRU cap instead of only being filtered on read. Added `test_committed_bookkeeping_does_not_outgrow_the_evicted_tree` to cover it.

File: docs/usage.md [hunk 4]
```diff
@@ -222,11 +223,76 @@ Each chat line is wrapped in the **same instruction template the model was train
...
+An annotated transcript — commit, branch, rewind, and a *different* continuation from the same
+prefix (toy scale, greedy, no weights, so the text itself is gibberish):
+
+```
+$ printf 'the cat\n/tree\nsat down\n/rewind 1\nran away\n/tree\n/rewind 99\n/frobnicate\n/quit\n' \
+    | .venv/bin/python scripts/generate.py --config config/toy.yaml --byte-fallback \
+        --interactive --temperature 0 --max-new-tokens 20 --rewind-depth 8
+
+rewind: up to 8 retained turn boundaries, at most 155648 bytes (8 x 19456 B/snapshot).
+>>> the cat        -> tttttttttttttttttttt      [committed #1; depth 2, 2 retained]
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply docs/usage.md]: ✅ Fixed in 5611219 — the annotated transcript now shows the real `{retained}/{max_depth} retained` commit-line format (verified against actual REPL output) instead of the abbreviated `N retained`.

File: docs/usage.md [hunk 5]
```diff
@@ -222,11 +223,76 @@ Each chat line is wrapped in the **same instruction template the model was train
...
+byte-for-byte under `--temperature 0`; that identity is what
+`tests/test_rewind_entry_point.py::test_rewound_branch_matches_direct_generation` asserts.
+
+**Memory.** Snapshots are the fixed-size recurrent state, so the tree's ceiling is flat
+arithmetic: `--rewind-depth x per_session_state_bytes(config)`, printed at startup. At toy
+scale that is `8 x 19456 B = 152 KiB`; at `config/poc.yaml` one snapshot is
+`n_layers x ((d_conv-1) x d_inner + n_heads x head_dim x d_state) x 4 B`
+(fp32-charged, the conservative direction), and the default depth 32 caps the tree at 32 of
+them. Nothing grows with conversation length — that is the whole point of an SSM state, and why
+`RewindTree` needs no paged-attention or radix-cache machinery.
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply docs/usage.md]: ✅ Fixed in 5611219 — reworded to scope the flat-memory claim to the retained *state snapshots* specifically, and to note the REPL's own transcript/bookkeeping (now pruned and bounded, per the two fixes above, but not literally nothing).

**Comment by travisgalloway** [comment]
Addressed review feedback:

- @copilot-pull-request-reviewer: all 5 line comments from the review fixed in 5611219 — pruned the REPL transcript dict and `SessionHistory._committed` to their trees' retained ids so both stay bounded by `--rewind-depth` instead of growing unbounded; fixed `/rewind N` (N>1) at the root to report the precise over-deep-rewind error instead of the generic "session is empty" message; and corrected the docs' annotated transcript + the "Nothing grows with conversation length" wording. Regression tests added for the two bookkeeping fixes; `test_state_independent_model_makes_the_identity_vacuous` is untouched.
